lib/worker/exec.go

```diff
@@ -0,0 +1,129 @@
+package worker
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+// Exec represents something to be executed in the Worker
+// TODO: I might just remove this interface in the next
+// PR. I introduced it thinking it could be useful for
+// the future cgroup implementation but now am not so
+// sure.
+type Exec interface {
+	// Command is the command for the exec.
+	Command() string
+	// Start starts the exec.
+	Start() error
+	// Stop stops the exec.
+	Stop() error
+	// PID returns the pid of the process once it has been started.
+	// Otherwise, and invalid pid 0 is returned.
+	PID() int
+	// Status returns the ProcessStatus struct corresponding
+	// to the started exec.
+	Status() (*ProcessStatus, error)
+}
+
+// process is a process executed or to be executed by the worker
+// TODO: Find a better name for this as the word "process" implies
+// that this it is already running.
+type process struct {
+	command    string
+	cmd        *exec.Cmd
+	startedBy  string
+	startedAt  time.Time
+	finishedAt time.Time // TODO: implement finished at
+}
+
+func newProcess(command string, cmd *exec.Cmd, startedBy string) *process {
+	return &process{
+		command:   command,
+		cmd:       cmd,
+		startedBy: startedBy,
+	}
+}
+
+// ProcessStatus represents a snapshot of a process status
+type ProcessStatus struct {
+	PID        int
+	StartedBy  string
+	State      string
+	StartedAt  time.Time
+	FinishedAt time.Time // TODO: implement finished at
+}
+
+func (p *process) Start() error {
+	p.startedAt = time.Now()
+	err := p.cmd.Start()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *process) Stop() error {
+	if p.cmd.Process == nil {
+		return fmt.Errorf("process not started")
+	}
+	return p.cmd.Process.Kill()
+}
+
+func (p *process) Command() string {
+	return p.command
+}
+
+func (p *process) PID() int {
+	if p.cmd.Process == nil {
+		return 0
+	}
+	return p.cmd.Process.Pid
+}
+
+func (p *process) Status() (*ProcessStatus, error) {
+	if p.cmd.Process == nil {
+		return nil, fmt.Errorf("process not started")
+	}
+	pid := p.PID()
```

Comment:
There's no guarantee that any of this data is consistent, as it could all change while it's being gathered. As-is, it's not _quite_ a race condition, but it's definitely one waiting to happen. As soon as something starts making decisions based on the resulting `ProcessStatus`, things will get dicey.

Reply:
`ProcessStatus` can never be consistent though right? Anything could happen in the background while the caller is getting an instance of this struct. For example, a memory exhaustion could get the process killed just after we determined the process state. `ProcessStatus` was intended as more of a "point in time" state.

An improvement could be to instead expose a function:

```
func (p *Process) DoSomethingWithProcessState(state *ProcessState) {
    p.mu.Lock()
    defer p.mu.Unlock()
}
```

which is mutually exclusive with the other mutating `Process` functions (Start, Stop, etc). This way we would at least know that the worker itself is not the one changing the state of the process.

Reply:
I've introduced a function like what I suggested above. It is not really used for anything right now but it could potentially give a way to look at a more consistent status (at least unaffected by the worker itself). I am worried that something like this could easily be misused to accidentally block all process operations for a long time (unless we also have a timeout). So, am still not sure of it's usefulness.